Foundry compiler + make build-foundry command

danilowhk opened the pull request:
- Added foundry compiler for complex solidity files
- Added "make build-foundry" command

(Have added StarkEx and UniswapV2 contracts)

## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Resolves #381

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Other information

<!-- Give an estimate of the time you spent on this PR in terms of work days. Did you spend 0.5 days on this PR or rather 2 days?  -->
Time spent on this PR:  

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

ClementWalter reviewed (commented):
on-the-fly comments:

- add `cache` folder to `.gitignore`
- why not putting the `foundry.toml` at the root of the project?
- we could add a `solidity` (or `solidity_contracts` or even just `contracts`) folder at the root as well (your current `foundry_compiler/src`)
- case in folder naming is not consistent (3 names and 3 ≠ cases!); use snake_case plz
- let's put all the builds in `solidity/build` and `.gitignore` this folder as well. Compiled files can be generated automatically
- may be relevant to use `git submodules` as we have started to copy/paste significant pieces of code from other github repos https://git-scm.com/book/en/v2/Git-Tools-Submodules but maybe out of the scope of this PR
- can't we use foundry for the "single files"? can't we flatten this level, ie put the file directly at the root `./solidity`?
- how is `foundry` installed and available both for anyone working on the project and the CI (for compiling the files that won't be committed anymore)?

ClementWalter commented:
should we remove the `docker` related things as well?

danilowhk commented:
Ok! Will work on the points in the first comment.
About the second, I am not familiar with docker and I have seen that the compilation script is running on CI so maybe it can be in a separate PR?

danilowhk commented:
About the last 2 points in the first comment:

- Yes we can have all the files on the foundry, but as mentioned above didn't want to change anything and just add the foundry first
- I believe foundry is installed globally and basically you can just go and run "forge build". For example (For uniswap v2 you can git-clone and just run "forge build" even without having initialized or created a foundry dir)

ClementWalter commented:
so you'd need at least to update the `README.md` with a line saying that `forge` is required for tests and a link to forge installation page (hint: )

Concerning the CI, you can add an action that install forge, see https://github.com/foundry-rs/foundry-rust-template/actions/runs/3596764004/workflow:

```
      - name: Install Foundry
        uses: onbjerg/foundry-toolchain@v1
        with:
          version: nightly
```

we can move this to another PR agreed

danilowhk commented:
- added cache folder to .gitignore
- changed foundry root to project root
- changed folder name
- added all the builds to .gitignore
- Added details on readme

TODO: 
- update submodules as specified
- migrate all the single files to foundry aswell
(as discussed I think this could be a separate PR)

ClementWalter commented on README.md:
````diff
@@ -156,6 +156,21 @@ To contribute, please check out [the contribution guide](./docs/CONTRIBUTING.md)
 ```
 
+## Install foundry
+To be able to verify and compare the abi and bytecode of the test contracts, first make sure you have foundry installed:
````
"To compile" ? we don't do any "verification" right?

danilowhk replied:
What do you mean by verification? As it is, if you run make build-foundry everything should compile

ClementWalter replied:
I don't know, you wrote "to verify and compare", and I was also wondering what you meant by this.
We'll foundry to compile AFAIK, not to do any verification

danilowhk replied:
Oh I mean manually verify the ABIs and Bycodes that are being used at the test/integration folder

ClementWalter reviewed (dismissed):
why is `StarkEx` inside a sub `src` folder and not the others?

danilowhk commented:
About StarkExt there is no particular reason it was just as it was when git cloned, should I change it?

ClementWalter commented:
> About StarkExt there is no particular reason it was just as it was when git cloned, should I change it?

would be more consistent with the other folders, but on the other hand I don't know how we'll do with the build artifacts more globally so you can leave it like that I guess

danilowhk commented:
Its easy to change, we can do it anytime we decide to leave everything as a pattern

danilowhk commented:
- Removed src from StarkEx and updated makefile
